### PR TITLE
WIP: Wrap tmc-core calls to BgTask and CancellableCallable to maintain current behaviour.

### PR DESCRIPTION
**Pull request by jamo**
More stuff to do, but a few tasks already migrated

### CODE REVIEW

**Comment by jamo** [comment]
Almost done, all tested with a simple java exercise. Let's try to get this merged and later cleanup some of the uglier code.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java [hunk 1]
```diff
@@ -76,118 +78,80 @@ public CheckForNewExercisesOrUpdates(boolean beQuiet, boolean backgroundCheck) {
 
     @Override
     public void actionPerformed(ActionEvent e) {
+        logger.info("action performed");
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java]: Should we log the name of the action too, or does to logging framework already log the location of the call?

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java [hunk 2]
```diff
@@ -76,118 +78,80 @@ public CheckForNewExercisesOrUpdates(boolean beQuiet, boolean backgroundCheck) {
...
+                                @Override
+                                public void actionPerformed(ActionEvent e) {
+                                    DownloadOrUpdateExercisesDialog.display(status.unlockable, status.downloadableUncompleted, status.updateable);
+                                }
+                            });
+                        } else {
+                            DownloadOrUpdateExercisesDialog.display(status.unlockable, status.downloadableUncompleted, status.updateable);
+                        }
+                    } else if (!beQuiet) {
+                        dialogs.displayMessage("No new exercises or updates to download.");
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java:119]: We should replace all hard coded UI strings with some i18n mechanism.

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java [hunk 3]
```diff
@@ -76,118 +78,80 @@ public CheckForNewExercisesOrUpdates(boolean beQuiet, boolean backgroundCheck) {
...
-                @Override
-                public void run() {
-                    lastAction.finish();
-                    if (!beQuiet || ex instanceof ObsoleteClientException) {
-                        dialogs.displayError("Failed to check for new exercises.\n" + ServerErrorHelper.getServerExceptionMsg(ex));
-                    }
-                }
+            @Override
+            public Course call() throws Exception {
+                logger.info("1Downloading course to refresh cache");
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java]: Why "1" ?

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java [hunk 4]
```diff
@@ -1,26 +1,28 @@
...
-import fi.helsinki.cs.tmc.core.exceptions.TmcCoreException;
-import java.lang.reflect.InvocationTargetException;
+import fi.helsinki.cs.tmc.utilities.AggregatingBgTaskListener;
+import fi.helsinki.cs.tmc.utilities.BgTask;
+import fi.helsinki.cs.tmc.utilities.BgTaskListener;
+import fi.helsinki.cs.tmc.utilities.CancellableCallable;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.ListenableFuture;
+import fi.helsinki.cs.tmc.model.CourseDb;
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java:19]: Wrong order for imports

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java [hunk 5]
```diff
@@ -44,106 +46,63 @@
...
+        this.settings = NbTmcSettings.getDefault();
     }
 
-    public void run() throws TmcCoreException {
-
-        int start = 0;
-        int end = Math.min(exercisesToDownload.size(), batchSize);
+    public void run() {
+//        final AggregatingBgTaskListener<TmcProjectInfo> aggregator
+//                = new AggregatingBgTaskListener<TmcProjectInfo>(exercisesToDownload.size(), whenAllDownloadsFinished);
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java]: Remove commented out code

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java [hunk 6]
```diff
@@ -44,106 +46,63 @@
...
-            try {
-                dlFuture = tmcCore.downloadExercises(batch);
-            } catch (TmcCoreException ex) {
-                List<Exercise> empty = new ArrayList<Exercise>();
-                dlFuture = Futures.immediateFuture(empty);
-                dialogs.displayError("Could not download exercises: " + batch, ex);
+    private BgTaskListener<Collection<Exercise>> whenDownloadsFinished = new BgTaskListener<Collection<Exercise>>() {
+        @Override
+        public void bgTaskReady(Collection<Exercise> exercises) {
+            logger.log(INFO, "1Opening projects.");
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java]: Extra "1"

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java [hunk 7]
```diff
@@ -44,106 +46,63 @@
...
+            List<TmcProjectInfo> projects = new ArrayList<TmcProjectInfo>(exercises.size());
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java:86]: Diamond operator

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java [hunk 8]
```diff
@@ -44,106 +46,63 @@
...
+            for (Exercise ex : exercises) {
+                projects.add(projectMediator.tryGetProjectForExercise(ex));
             }
-
-            Futures.addCallback(dlFuture, new ProjectOpener(exerciseDownload, end));
+            projectMediator.openProjects(projects);
+            logger.log(INFO, "2Opening projects.");
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/DownloadExercisesAction.java]: "2"

File: tmc-plugin/src/fi/helsinki/cs/tmc/ui/DownloadOrUpdateExercisesDialog.java [hunk 9]
```diff
@@ -126,14 +126,14 @@ private void updateSelectAllButtonState() {
...
-        } catch (TmcCoreException ex) {
-            Exceptions.printStackTrace(ex);
-            ConvenientDialogDisplayer.getDefault()
-                .displayError("Failed to download exercises.\n" + ServerErrorHelper.getServerExceptionMsg(ex));
-        }
+//        } catch (TmcCoreException ex) {
+//            Exceptions.printStackTrace(ex);
+//            ConvenientDialogDisplayer.getDefault()
+//                .displayError("Failed to download exercises.\n" + ServerErrorHelper.getServerExceptionMsg(ex));
+//        }
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/ui/DownloadOrUpdateExercisesDialog.java]: Commented out code in repo

File: tmc-plugin/src/fi/helsinki/cs/tmc/ui/PreferencesPanel.java [hunk 10]
```diff
@@ -60,11 +61,10 @@ public boolean noneIsEmpty() {
...
             if (obj instanceof RefreshSettings) {
-                RefreshSettings that = (RefreshSettings)obj;
-                return
-                        ObjectUtils.equals(this.username, that.username) &&
-                        ObjectUtils.equals(this.password, that.password) &&
-                        ObjectUtils.equals(this.baseUrl, that.baseUrl);
+                RefreshSettings that = (RefreshSettings) obj;
+                return ObjectUtils.equals(this.username, that.username)
+                        && ObjectUtils.equals(this.password, that.password)
+                        && ObjectUtils.equals(this.baseUrl, that.baseUrl);
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/ui/PreferencesPanel.java:67]: Replace 3rd party utils with standard lib:

http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#equals(java.lang.Object,%20java.lang.Object)

File: tmc-plugin/src/fi/helsinki/cs/tmc/ui/PreferencesPanel.java [hunk 11]
```diff
@@ -183,7 +183,7 @@ public void run() {
     @Override
     public void setSelectedCourseName(String courseName) {
         for (int i = 0; i < coursesComboBox.getItemCount(); ++i) {
-            if (((Course)coursesComboBox.getItemAt(i)).getName().equals(courseName)) {
+            if (((Course) coursesComboBox.getItemAt(i)).getName().equals(courseName)) {
```
Comment (ljleppan) [inline tmc-plugin/src/fi/helsinki/cs/tmc/ui/PreferencesPanel.java:186]: An NPE from this line will be really hard to debug, split into multiple lines?

File: tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java (hunk 2 above)
Comment (jamo) [inline tmc-plugin/src/fi/helsinki/cs/tmc/actions/CheckForNewExercisesOrUpdates.java:119]: Yes, we should

**Pull request merged**